core/store.go

```diff
@@ -290,6 +291,111 @@ func Get(k string) *Obj {
 	return getHelper(k, true)
 }
 
+var (
+	keyList []string
```

Comment:
This is a global variable. What happens if multiple scan commands are issued from different clients around the same time?

Lets take an example scenario:
Client 1 issues SCAN command:
```
redis> SCAN 0
1) "2"
2) 1) "key1"
   2) "key2"
```

Now before client 1 issues `SCAN 2` client 2 issues the command `SCAN 0`.
What is the expected behaviour in this case?
How should the DiceDB behave in this scenario?

Reply:
Yes currently this works fine with single client only because the cursor is maintained globally, so I think I can maintain a state for each client if it's feasible and make keylist specific to client and not global. Is this fair or can you give me suggestions here.

Reply:
@rohitlohar45 I feel implementation of the SCAN complicated is complicated compared to other commands. We need to really think this through.

I am bit help up with other tasks. Lets discuss this early next week.

Reply:
I agree  with @soumya-codes .
the current implementation has also one problem.
The reason is you are initialising each time keyList from keypool which is map in which order is not guaranteed,
and you are continuing the iteration from the cursor value.

here i dont get key5 in full iteration. which is present before iteration started and was present during full iteration. 
![Screenshot from 2024-08-23 18-53-25](https://github.com/user-attachments/assets/7920c5a6-c74e-4af3-9391-5da664dc921b)

 

> A full iteration always retrieves all the elements that were present in the collection from the start to the end of a full iteration. This means that if a given element is inside the collection when an iteration is started, and is still there when an iteration terminates, then at some point SCAN returned it to the user.
 
 Let me know if i misunderstood the doc.

Reply:
fyi @rohitlohar45

Reply:
@soumya-codes yes.
This is not simple feature.